StreamFlix Content Analytics — Phase 1: Data Cleaning & Quality Checks

In [1]:
import pandas as pd
from sqlalchemy import create_engine

password = "mayuri27"
engine = create_engine(f"mysql+pymysql://root:{password}@localhost:3306/streamflix")

## Check 1: Churn Date Validation

Verifies that every churned subscriber's churn date falls after their signup date.

In [2]:
query1 = """
SELECT subscriber_id, signup_date, churn_date
FROM subscribers
WHERE is_active = FALSE
AND churn_date IS NOT NULL
AND churn_date <= signup_date;
"""

result1 = pd.read_sql_query(query1, engine)
print(f"Rows flagged: {len(result1)}")
result1

Rows flagged: 0


,subscriber_id,signup_date,churn_date


## Check 2: Active Subscribers Should Have No Churn Date

Verifies that subscribers currently marked active do not have a churn date recorded.

In [3]:
query2 = """
SELECT subscriber_id, is_active, churn_date
FROM subscribers
WHERE is_active = TRUE
AND churn_date IS NOT NULL;
"""

result2 = pd.read_sql_query(query2, engine)
print(f"Rows flagged: {len(result2)}")
result2

Rows flagged: 0


,subscriber_id,is_active,churn_date


## Check 2b: Inactive Subscribers Should Have a Churn Date

Verifies the reverse case — every inactive subscriber has a churn date on record.

In [4]:
query2b = """
SELECT subscriber_id, is_active, churn_date
FROM subscribers
WHERE is_active = FALSE
AND churn_date IS NULL;
"""

result2b = pd.read_sql_query(query2b, engine)
print(f"Rows flagged: {len(result2b)}")
result2b

Rows flagged: 0


,subscriber_id,is_active,churn_date


## Check 3: Completion Percentage Consistency

Verifies that completion_pct is mathematically consistent with watch_duration_min divided by content_duration_min.

In [5]:
query3 = """
SELECT watch_id, subscriber_id, title_id,
content_duration_min, watch_duration_min, completion_pct,
ROUND(100.0 * watch_duration_min / NULLIF(content_duration_min, 0), 1) AS calculated_pct,
ROUND(completion_pct - (100.0 * watch_duration_min / NULLIF(content_duration_min, 0)), 1) AS diff
FROM watch_history
WHERE ABS(completion_pct - (100.0 * watch_duration_min / NULLIF(content_duration_min, 0))) > 2
ORDER BY ABS(diff) DESC;
"""

result3 = pd.read_sql_query(query3, engine)
print(f"Rows flagged: {len(result3)}")
result3

Rows flagged: 0


,watch_id,subscriber_id,title_id,content_duration_min,watch_duration_min,completion_pct,calculated_pct,diff


## Check 4: Valid Sentiment Values

Verifies that every review's sentiment value is one of the three expected categories.

In [6]:
query4 = """
SELECT DISTINCT sentiment
FROM reviews
WHERE sentiment NOT IN ('Positive', 'Neutral', 'Negative');
"""

result4 = pd.read_sql_query(query4, engine)
print(f"Rows flagged: {len(result4)}")
result4

Rows flagged: 0


,sentiment


## Check 5a: Watchlist References a Valid Subscriber

Verifies that every watchlist entry points to a subscriber that actually exists.

In [7]:
query5a = """
SELECT wl.watchlist_id, wl.subscriber_id
FROM watchlist wl
LEFT JOIN subscribers s ON s.subscriber_id = wl.subscriber_id
WHERE s.subscriber_id IS NULL;
"""

result5a = pd.read_sql_query(query5a, engine)
print(f"Rows flagged: {len(result5a)}")
result5a

Rows flagged: 0


,watchlist_id,subscriber_id


## Check 5b: Watchlist References a Valid Title

Verifies the same for titles — every watchlist entry must point to a real title.

In [8]:
query5b = """
SELECT wl.watchlist_id, wl.title_id
FROM watchlist wl
LEFT JOIN titles t ON t.title_id = wl.title_id
WHERE t.title_id IS NULL;
"""

result5b = pd.read_sql_query(query5b, engine)
print(f"Rows flagged: {len(result5b)}")
result5b

Rows flagged: 0


,watchlist_id,title_id


## Check 6: Short Tenure but Active (Informational)

Flags subscribers with one month or less of tenure who are still marked active. This is expected, not an error.

In [9]:
query6 = """
SELECT subscriber_id, signup_date, tenure_months, is_active
FROM subscribers
WHERE is_active = TRUE
AND tenure_months <= 1
ORDER BY tenure_months;
"""

result6 = pd.read_sql_query(query6, engine)
print(f"Rows flagged: {len(result6)}")
result6

Rows flagged: 321


,subscriber_id,signup_date,tenure_months,is_active
0,SUB100066,2026-05-18,1,1
1,SUB100102,2026-04-30,1,1
2,SUB100144,2026-04-10,1,1
3,SUB100170,2026-04-14,1,1
4,SUB100270,2026-05-05,1,1
...,...,...,...,...
316,SUB114901,2026-05-19,1,1
317,SUB114918,2026-04-22,1,1
318,SUB114941,2026-05-13,1,1
319,SUB114943,2026-04-29,1,1


## Overall Summary

All six data quality checks passed with zero issues found. The only flagged result was informational (Check 6) — 321 recently-signed-up subscribers still marked active, which is expected behavior rather than a data problem. The StreamFlix dataset shows strong referential integrity, consistent date logic, valid categorical values, and accurate calculated fields.